main/src/main/scala/sbt/Extracted.scala

```diff
@@ -120,10 +120,22 @@ final case class Extracted(structure: BuildStructure,
     structure.data.get(scope, key) getOrElse sys.error(
       display.show(ScopedKey(scope, key)) + " is undefined.")
 
-  def append(settings: Seq[Setting[_]], state: State): State = {
+  /** Appends the given settings to the original build state settings. */
```

Comment:
I suggest adding information about the gotcha of discarding session settings, perhaps marking this as deprecated

Reply:
I think being more explicit about session settings makes sense.

But if we're assuming that discarding them is intended behaviour and a valid use case, we can't mark this as deprecated.

Reply:
It wasn't exactly intended behavior, at least originally. The method `append`, with the current semantics, should have been deprecated long ago. As a reference:

On 06/03/13 02:27, Mark Harrah wrote:
> On 06/03/13 01:34, Antonio Cunei wrote:
>> However, I guess Extracted.append() could mirror more closely the
>> operation of the set command, otherwise the result may be surprising to
>> users. (Also, an equivalent to "set every" might be handy). Is this
>> analysis correct, or is there a different recommended way to patch the
>> session?
>
> Your understanding is correct.  There are different semantics that are useful at different times.  I'm not real sure if it is the right thing for commands to effectively do the same as set.  They don't tend to be saveable since there isn't a corresponding source.  One way forward is to deprecate append and replace it with two (or more) methods with documented semantics.

Reply:
Maybe we should deprecate in favour of `appendWithSession` and `appendDiscardingSession`?

Reply:
Not that historical text by Mark should be taken as sacred, if I am reading this right he's saying that _while he was unsure if it's the right thing_, he was discarding intentionally and thus suggesting we have "two (or more) methods".

Here's my suggestion:
1. Add `appendWithoutSession(settings: Seq[Setting[_]], state: State)` and `appendWithSession(settings: Seq[Setting[_]], state: State)`.
2. Deprecate `append(...)` explaining that the `append` discards session, and that the user is encourage to move to either `appendWithSession(...)` or `appendWithoutSession(...)`.

I am also tempted to add an alias called `set(...)` for `appendWithSession(...)` as it would have the same semantics as the `set` command, but I'm afraid `append` vs `set` would be too subtle.